### PR TITLE
Connect positive motifs to pipeline

### PR DESCRIPTION
**Pull request by yael1994**

### CODE REVIEW

File: IgOmeProfiling_pipeline.py [hunk 1]
```diff
@@ -180,7 +182,16 @@ def run_pipeline(fastq_path, barcode2samplename_path, samplename2biologicalcondi
...
+                             '--use_positive_motifs_script' if use_positive_motifs_script else '',
+                             '' if invalid_mix is None else f'--invalid_mix {invalid_mix}',
+                             '' if threshold_mean is None else f'--threshold_mean {threshold_mean}',
+                             '' if threshold_std is None else f'--threshold_std {threshold_std}',
+                             '' if threshold_median is None else f'--threshold_median {threshold_median}',
+                             f'--min_max_difference' if min_max_difference else '',
+                             f'--normalize_factor {normalize_factor}', f'--normalize_method_hits {normalize_method_hits}',
+                             f'--normalize_section {normalize_section}',
+                             '' if fixed_min is None else f'--fixed_min {fixed_min}',
+                             '' if fixed_max is None else f'--fixed_max {fixed_max}']
```
Comment (shacharmo) [inline IgOmeProfiling_pipeline.py]: Argument names are not clear enugh, it's not clear they are part of positive motifs filter.
Also, change `use_positive_motifs_script` to `filter_positive_motifs`

File: IgOmeProfiling_pipeline.py [hunk 2]
```diff
@@ -180,7 +182,16 @@ def run_pipeline(fastq_path, barcode2samplename_path, samplename2biologicalcondi
                              '--no_rpm_factor' if no_rpm_factor else '',
                              '--no_use_rpm_faa_scanning' if no_use_rpm_faa_scanning else '',
                              f'--no_output_sequences_scanning' if no_output_sequences_scanning else '',
-                             f'--error_path {error_path}', '-v' if verbose else '', f'-q {queue}','-m' if use_mapitope else '']
```
Comment (shacharmo) [inline IgOmeProfiling_pipeline.py:183]: Did you mean to delete to row? It seems script needs these arguments

File: IgOmeProfiling_pipeline.py [hunk 3]
```diff
@@ -300,6 +311,27 @@ def run_pipeline(fastq_path, barcode2samplename_path, samplename2biologicalcondi
     parser.add_argument('--no_rpm_factor', action='store_true', help='Disable multiplication hits by factor rpm for normalization')
     parser.add_argument('--no_use_rpm_faa_scanning', action='store_true', help='Disable performance of scanning script with unique rpm faa file')
     parser.add_argument('--no_output_sequences_scanning', action='store_true', help='Disable storing the output sequences that had hits')
+    parser.add_argument('--use_positive_motifs_script', action='store_true', help='Before classification model, remain only positive motifs from the csv file')
```
Comment (shacharmo) [inline IgOmeProfiling_pipeline.py]: Change name as in comment above.
Help should be "filter only positive motifs. This is done after motifs generation and before ranking"

File: IgOmeProfiling_pipeline.py [hunk 4]
```diff
@@ -300,6 +311,27 @@ def run_pipeline(fastq_path, barcode2samplename_path, samplename2biologicalcondi
     parser.add_argument('--no_rpm_factor', action='store_true', help='Disable multiplication hits by factor rpm for normalization')
     parser.add_argument('--no_use_rpm_faa_scanning', action='store_true', help='Disable performance of scanning script with unique rpm faa file')
     parser.add_argument('--no_output_sequences_scanning', action='store_true', help='Disable storing the output sequences that had hits')
+    parser.add_argument('--use_positive_motifs_script', action='store_true', help='Before classification model, remain only positive motifs from the csv file')
+    parser.add_argument('--invalid_mix',type=str, default=None, help='A argument to know if there is compare to naive')
```
Comment (shacharmo) [inline IgOmeProfiling_pipeline.py]: Shouldn't the help be:
Sample name considered negative. e.g. "native"
?

Should default be "naive"?

File: IgOmeProfiling_pipeline.py [hunk 5]
```diff
@@ -300,6 +311,27 @@ def run_pipeline(fastq_path, barcode2samplename_path, samplename2biologicalcondi
     parser.add_argument('--no_rpm_factor', action='store_true', help='Disable multiplication hits by factor rpm for normalization')
     parser.add_argument('--no_use_rpm_faa_scanning', action='store_true', help='Disable performance of scanning script with unique rpm faa file')
     parser.add_argument('--no_output_sequences_scanning', action='store_true', help='Disable storing the output sequences that had hits')
+    parser.add_argument('--use_positive_motifs_script', action='store_true', help='Before classification model, remain only positive motifs from the csv file')
+    parser.add_argument('--invalid_mix',type=str, default=None, help='A argument to know if there is compare to naive')
+    parser.add_argument('--threshold_mean', default=None,
```
Comment (shacharmo) [inline IgOmeProfiling_pipeline.py]: Help should be:
Positive motif threshold of the difference between the mean of the BC and mean of others

Fix other arguments helps to be similar

File: IgOmeProfiling_pipeline.py [hunk 6]
```diff
@@ -300,6 +311,27 @@ def run_pipeline(fastq_path, barcode2samplename_path, samplename2biologicalcondi
...
+                        type=lambda x: float(x) if 0 < float(x) < 1 
+                        else parser.error(f'The threshold of the std diffrence should be between 0 to 1'),
+                        help='If the diffrenece between the std of BC to the std of other is bigger than the motif is seperate')
+    parser.add_argument('--threshold_median', default=None,
+                        type=lambda x: float(x) if 0 < float(x) < 1 
+                        else parser.error(f'The threshold of the median diffrence should be between 0 to 1'),
+                        help='If the diffrenece between the median of BC to the median of other is bigger than the motif is seperate')
+    parser.add_argument('--min_max_difference', action='store_true', help='motifs is positive if the minmal val of bc is bigger than the maximal value of other')
+    parser.add_argument('--normalize_factor', choices=['linear', 'log'], default='linear', help='Type of factor on number for highlight them')
+    parser.add_argument('--normalize_method_hits', choices=['min_max', 'max', 'fixed_min_max'], default='min_max',
```
Comment (shacharmo) [inline IgOmeProfiling_pipeline.py]: This doesn't explain what each method does

File: IgOmeProfiling_pipeline.py [hunk 7]
```diff
@@ -300,6 +311,27 @@ def run_pipeline(fastq_path, barcode2samplename_path, samplename2biologicalcondi
...
+                        help='If the diffrenece between the median of BC to the median of other is bigger than the motif is seperate')
+    parser.add_argument('--min_max_difference', action='store_true', help='motifs is positive if the minmal val of bc is bigger than the maximal value of other')
+    parser.add_argument('--normalize_factor', choices=['linear', 'log'], default='linear', help='Type of factor on number for highlight them')
+    parser.add_argument('--normalize_method_hits', choices=['min_max', 'max', 'fixed_min_max'], default='min_max', 
+                        help='Type of method to do the normaliztion on hits data, change the values to be between 0 to 1')
+    parser.add_argument('--normalize_section', choices=['per_motif','per_exp'],  default='per_motif', help='Calculate the min and max per motifs or over all the exp data')
+    parser.add_argument('--fixed_min', type=int, default=None, help='In case of fixed_min_max for normalize_method_hits set the minimum value')
+    parser.add_argument('--fixed_max', type=int, default=None, help='In case of fixed_min_max for normalize_method_hits set the maximum value')
```
Comment (shacharmo) [inline IgOmeProfiling_pipeline.py]: Is there validation if arguments not set but `fixed_min_max` is set?

File: model_fitting/positive_motifs.py [hunk 8]
```diff
@@ -105,27 +105,36 @@ def calculation(df, label):
...
-            and (threshold_std is None or df.loc['std_BC', motif_name] - df.loc['std_other', motif_name] > threshold_std) \
-            and (threshold_median is None or df.loc['median_BC', motif_name] - df.loc['median_other', motif_name] > threshold_median) \
-            and (not min_max_difference or df.loc['min_BC', motif_name] > df.loc['max_other', motif_name]):
-            positive_motifs.append(motif_name)
-            motifs_value.append('positive')
+        if rank_method == 'hits' and is_rpm_normalize:
+            if (threshold_mean is None or (df.loc['mean_BC', motif_name] - df.loc['mean_other', motif_name]) / df.loc['mean_BC', motif_name] > threshold_mean) \
+                and (threshold_std is None or (df.loc['std_BC', motif_name] - df.loc['std_other', motif_name]) / df.loc['std_BC', motif_name] > threshold_std) \
+                and (threshold_median is None or (df.loc['median_BC', motif_name] - df.loc['median_other', motif_name]) / df.loc['median_BC', motif_name] > threshold_median) \
+                and (not min_max_difference or df.loc['min_BC', motif_name] > df.loc['max_other', motif_name]):
```
Comment (shacharmo) [inline model_fitting/positive_motifs.py]: Refactor to method if `is_positive` to avoid duplicate code

**Review by shacharmo — CHANGES_REQUESTED** [review]

File: IgOmeProfiling_pipeline.py (hunk 4 above)
Comment (yael1994) [reply IgOmeProfiling_pipeline.py]: Not all the time there are naive samples.

File: IgOmeProfiling_pipeline.py (hunk 7 above)
Comment (yael1994) [reply IgOmeProfiling_pipeline.py]: There is a validation in the script positive motifs - 
```
    if normalize_method_hits =='fixed_min_max':
        if fixed_max is None:
            fixed_max = max_exp
        if fixed_min is None:
            fixed_min = 0
```

**Review by shacharmo — APPROVED** [review]

**Pull request merged**